### ISSUE TITLE
[BUG] Data Race Issues in dd-trace-go with go-redis/v9: Concurrent Read/Write Access Detected

### ISSUE THREAD

**Issue by RyoMiyashita** [DataDog/dd-trace-go#2376]
Hello, I have updated to `go-redis/v9` and CI detected a data race.

**Version of dd-trace-go**

v1.58.0

**Describe what happened:**

A data race condition was detected when using `dd-trace-go` with the `go-redis/v9` package. This involves concurrent read and write operations in the datadogHook and internal pool methods of `go-redis/v9`. The race condition occurs during the execution of WrapClient or NewClient methods from `dd-trace-go/v1/contrib/redis/go-redis.v9`.

**Describe what you expected:**
The expected behavior is the safe and race-condition-free execution of `dd-trace-go` methods when interfacing with `go-redis/v9`, ensuring thread safety and data integrity.

**Steps to reproduce the issue:**

- source code
```go
func main() {
	ctx := context.Background()
	opts, err := getOpts()
	if err != nil {
		panic(err)
	}
	client := redistrace.NewClient(opts, redistrace.WithAnalytics(true), redistrace.WithServiceName("redis"))

	client.Set(ctx, "key", "value", 0)

	var check string
	if err := client.Get(ctx, "key").Scan(&check); err != nil {
		panic(err)
	}

	fmt.Println("check:", check)
}

func getOpts() (*redis.Options, error) {
	redisURL := envutil.GetString("REDIS_URL", "redis://localhost:6379")
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, err
	}
	opts.MaxRetries = envutil.GetInt("REDIS_MAX_RETRIES", 1)
	opts.PoolSize = envutil.GetInt("REDIS_POOL_SIZE", 10*runtime.NumCPU())
	opts.MinIdleConns = envutil.GetInt("REDIS_MIN_IDLE_CONNS", runtime.NumCPU())
	opts.ConnMaxLifetime = envutil.GetDuration("REDIS_MAX_CONN_AGE", 30*time.Minute)
	opts.ReadTimeout = envutil.GetDuration("REDIS_READ_TIMEOUT", 2*time.Second)
	opts.WriteTimeout = envutil.GetDuration("REDIS_WRITE_TIMEOUT", 3*time.Second)
	opts.PoolTimeout = envutil.GetDuration("REDIS_POOL_TIMEOUT", 5*time.Second)
	return opts, nil
}
```
- run cmd
```
go run -race main.go 
```
- found 10 data races (only one is listed)
```
==================
WARNING: DATA RACE
Read at 0x00c00011cd20 by goroutine 12:
  runtime.slicecopy()
      /usr/local/opt/go/libexec/src/runtime/slice.go:310 +0x0
  gopkg.in/DataDog/dd-trace-go.v1/contrib/redis/go-redis%2ev9.(*datadogHook).DialHook.func1()
      /Users/ryo/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.58.0/contrib/redis/go-redis.v9/redis.go:118 +0x339
  github.com/redis/go-redis/v9.(*hooksMixin).dialHook()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/redis.go:168 +0xb0
  github.com/redis/go-redis/v9.(*hooksMixin).dialHook-fm()
      <autogenerated>:1 +0x29
  github.com/redis/go-redis/v9.newConnPool.func1()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/options.go:504 +0xa4
  github.com/redis/go-redis/v9/internal/pool.(*ConnPool).dialConn()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/internal/pool/pool.go:205 +0x186
  github.com/redis/go-redis/v9/internal/pool.(*ConnPool).addIdleConn()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/internal/pool/pool.go:143 +0x6f
  github.com/redis/go-redis/v9/internal/pool.(*ConnPool).checkMinIdleConns.func1()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/internal/pool/pool.go:126 +0x34

Previous write at 0x00c00011cd20 by main goroutine:
  gopkg.in/DataDog/dd-trace-go.v1/contrib/redis/go-redis%2ev9.additionalTagOptions()
      /Users/ryo/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.58.0/contrib/redis/go-redis.v9/redis.go:104 +0xe35
  gopkg.in/DataDog/dd-trace-go.v1/contrib/redis/go-redis%2ev9.WrapClient()
      /Users/ryo/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.58.0/contrib/redis/go-redis.v9/redis.go:60 +0xa4
  gopkg.in/DataDog/dd-trace-go.v1/contrib/redis/go-redis%2ev9.NewClient()
      /Users/ryo/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.58.0/contrib/redis/go-redis.v9/redis.go:46 +0x5e
  main.main()
      /Users/ryo/Repositories/dd-trace-test/main.go:21 +0x104

Goroutine 12 (running) created at:
  github.com/redis/go-redis/v9/internal/pool.(*ConnPool).checkMinIdleConns()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/internal/pool/pool.go:125 +0x73
  github.com/redis/go-redis/v9/internal/pool.NewConnPool()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/internal/pool/pool.go:109 +0x26e
  github.com/redis/go-redis/v9.newConnPool()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/options.go:502 +0x337
  github.com/redis/go-redis/v9.NewClient()
      /Users/ryo/go/pkg/mod/github.com/redis/go-redis/v9@v9.3.0/redis.go:623 +0x208
  gopkg.in/DataDog/dd-trace-go.v1/contrib/redis/go-redis%2ev9.NewClient()
      /Users/ryo/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.58.0/contrib/redis/go-redis.v9/redis.go:45 +0x3b
  main.main()
      /Users/ryo/Repositories/dd-trace-test/main.go:21 +0x104
==================
```

**Additional environment details (Version of Go, Operating System, etc.):**
- Go version: go1.20.5 darwin/amd64
- Operating System: macOS 13.6
- Version of go-redis/v9: v9.3.0

**Comment by RyoMiyashita** [comment]
I've made an interesting observation while troubleshooting the data race issue. When I set `MinIdleConns` to `0`, the data race doesn't seem to occur. However, it's important to note that this doesn't necessarily resolve the underlying problem. It seems more like a workaround or a temporary fix. We still need a proper solution that addresses the root cause of the data race, ensuring that it doesn't happen regardless of the `MinIdleConns` setting. Looking forward to further insights or fixes on this matter.
